### ISSUE TITLE
Docker start <container_name> returns "id already in use"

### ISSUE THREAD

**Issue by DoronCohen** [moby/moby#38346]
**Description**
`docker start redis-server` fails on this error message:

    Error response from daemon: id already in use
    Error: failed to start containers: redis-server

Notice I dont get "HASH is already in use" like in #36145. also the docker version here is 18.06
That container is stopped and not running. We run this command via SSH on an AWS instance which is stopped and started by demand.

**Steps to reproduce the issue:**
This bug happens occasionally so I didn't succeed reproducing it but the original flow was:
1. `docker run -d -v /redis/redis.conf:/usr/local/etc/redis/redis.conf --restart on-failure:3 --name redis-server --publish 6379:6379 redis:3.0 redis-server /usr/local/etc/redis/redis.conf`
2. stop redis container
3. stop AWS instance
4. start AWS instance
5. `docker start redis-server`

**Describe the results you received:**

    Error response from daemon: id already in use
    Error: failed to start containers: redis-server

**Describe the results you expected:**
running redis container

**Additional information you deem important (e.g. issue happens only occasionally):**
issue happens only occasionally

**Output of `docker version`:**

```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:58 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:24 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 3
Server Version: 18.06.1-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 23
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 apparmor
Kernel Version: 3.13.0-161-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 992.3MiB
Name: ip-172-31-11-76
ID: DG3V:VWIJ:KQ4X:W7X5:E5WR:RM6A:SQUE:EMG5:TT3K:MCHT:6SRX:PJXF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running on a `t2.micro` instance in AWS

**Comment by cpuguy83** [comment]
It seems like the container did not get cleaned up in containerd on the last exit for some reason and we  aren't handling that well here docker. 

I can make a PR to work around this issue which is probably good, but need to figure out how we got to that state.

Do you have any more details on the node that this was an issue on?

**Comment by cpuguy83** [comment]
I believe this should work around the issue: https://github.com/moby/moby/pull/38364

**Comment by DoronCohen** [comment]
Sorry but I can't provide anymore details because this is automated and it is not reproducible.
If I will get to connect to a live instance with this error what should I run on it and what details do you need ?

**Comment by cpuguy83** [comment]
More looking at how it might be reproduced, and I can't really think of how other than something else messing around with containerd, which doesn't seem all that plausible.

**Comment by cpuguy83** [comment]
Actually since this was created I saw another report of this, so perhaps we introduced a bug somewhere.

**Comment by DoronCohen** [comment]
ok I got a machine with this problem
daemon logs:

```
time="2018-12-27T10:56:59.068671675Z" level=warning msg="Error while setting daemon root propagation, this is not generally critical but may cause some functionality to not work or fallback to less desirable behavior" dir=/var/lib/docker error="error writing file to signal mount cleanup on shutdown: open /var/run/docker/unmount-on-shutdown: no such file or directory"
time="2018-12-27T10:56:59.081173578Z" level=info msg="libcontainerd: started new docker-containerd process" pid=1083
time="2018-12-27T10:56:59.081220523Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2018-12-27T10:56:59.081229171Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2018-12-27T10:56:59.081879566Z" level=info msg="ccResolverWrapper: sending new addresses to cc: [{unix:///var/run/docker/containerd/docker-containerd.sock 0  <nil>}]" module=grpc
time="2018-12-27T10:56:59.081901899Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2018-12-27T10:56:59.081953385Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc4203cb7f0, CONNECTING" module=grpc
Waiting for /var/run/docker.sock
/var/run/docker.sock is up
[34mINFO[0m[0000] starting containerd                           [34mrevision[0m=468a545b9edcd5932818eb9de8e72413e616e86e [34mversion[0m=v1.1.2
[34mINFO[0m[0000] loading plugin "io.containerd.content.v1.content"...  [34mtype[0m=io.containerd.content.v1
[34mINFO[0m[0000] loading plugin "io.containerd.snapshotter.v1.btrfs"...  [34mtype[0m=io.containerd.snapshotter.v1
[33mWARN[0m[0000] failed to load plugin io.containerd.snapshotter.v1.btrfs  [33merror[0m="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter"
[34mINFO[0m[0000] loading plugin "io.containerd.snapshotter.v1.aufs"...  [34mtype[0m=io.containerd.snapshotter.v1
[34mINFO[0m[0000] loading plugin "io.containerd.snapshotter.v1.native"...  [34mtype[0m=io.containerd.snapshotter.v1
[34mINFO[0m[0000] loading plugin "io.containerd.snapshotter.v1.overlayfs"...  [34mtype[0m=io.containerd.snapshotter.v1
[34mINFO[0m[0000] loading plugin "io.containerd.snapshotter.v1.zfs"...  [34mtype[0m=io.containerd.snapshotter.v1
[33mWARN[0m[0000] failed to load plugin io.containerd.snapshotter.v1.zfs  [33merror[0m="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.zfs must be a zfs filesystem to be used with the zfs snapshotter"
[34mINFO[0m[0000] loading plugin "io.containerd.metadata.v1.bolt"...  [34mtype[0m=io.containerd.metadata.v1
[33mWARN[0m[0000] could not use snapshotter btrfs in metadata plugin  [33merror[0m="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.btrfs must be a btrfs filesystem to be used with the btrfs snapshotter"
[33mWARN[0m[0000] could not use snapshotter zfs in metadata plugin  [33merror[0m="path /var/lib/docker/containerd/daemon/io.containerd.snapshotter.v1.zfs must be a zfs filesystem to be used with the zfs snapshotter"
[34mINFO[0m[0000] loading plugin "io.containerd.differ.v1.walking"...  [34mtype[0m=io.containerd.differ.v1
[34mINFO[0m[0000] loading plugin "io.containerd.gc.v1.scheduler"...  [34mtype[0m=io.containerd.gc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.containers-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.content-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.diff-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.images-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.leases-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.namespaces-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.snapshots-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.monitor.v1.cgroups"...  [34mtype[0m=io.containerd.monitor.v1
[34mINFO[0m[0000] loading plugin "io.containerd.runtime.v1.linux"...  [34mtype[0m=io.containerd.runtime.v1
[34mINFO[0m[0000] loading plugin "io.containerd.service.v1.tasks-service"...  [34mtype[0m=io.containerd.service.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.containers"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.content"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.diff"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.events"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.healthcheck"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.images"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.leases"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.namespaces"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.snapshots"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.tasks"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.version"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] loading plugin "io.containerd.grpc.v1.introspection"...  [34mtype[0m=io.containerd.grpc.v1
[34mINFO[0m[0000] serving...                                    [34maddress[0m="/var/run/docker/containerd/docker-containerd-debug.sock"
[34mINFO[0m[0000] serving...                                    [34maddress[0m="/var/run/docker/containerd/docker-containerd.sock"
[34mINFO[0m[0000] containerd successfully booted in 0.041631s  
time="2018-12-27T10:56:59.550311664Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc4203cb7f0, READY" module=grpc
time="2018-12-27T10:56:59.615946110Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2018-12-27T10:56:59.616124899Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2018-12-27T10:56:59.616211381Z" level=info msg="ccResolverWrapper: sending new addresses to cc: [{unix:///var/run/docker/containerd/docker-containerd.sock 0  <nil>}]" module=grpc
time="2018-12-27T10:56:59.616305358Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2018-12-27T10:56:59.616390543Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc420186df0, CONNECTING" module=grpc
time="2018-12-27T10:56:59.616875770Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc420186df0, READY" module=grpc
time="2018-12-27T10:56:59.632320636Z" level=info msg="[graphdriver] using prior storage driver: aufs"
time="2018-12-27T10:56:59.703873384Z" level=info msg="Graph migration to content-addressability took 0.00 seconds"
time="2018-12-27T10:56:59.704050862Z" level=warning msg="Your kernel does not support swap memory limit"
time="2018-12-27T10:56:59.704084416Z" level=warning msg="Your kernel does not support cgroup rt period"
time="2018-12-27T10:56:59.704095478Z" level=warning msg="Your kernel does not support cgroup rt runtime"
time="2018-12-27T10:56:59.704189550Z" level=warning msg="mountpoint for pids not found"
time="2018-12-27T10:56:59.704533301Z" level=info msg="parsed scheme: \"unix\"" module=grpc
time="2018-12-27T10:56:59.704606261Z" level=info msg="scheme \"unix\" not registered, fallback to default scheme" module=grpc
time="2018-12-27T10:56:59.704709941Z" level=info msg="ccResolverWrapper: sending new addresses to cc: [{unix:///var/run/docker/containerd/docker-containerd.sock 0  <nil>}]" module=grpc
time="2018-12-27T10:56:59.704776374Z" level=info msg="ClientConn switching balancer to \"pick_first\"" module=grpc
time="2018-12-27T10:56:59.704883011Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc4201c44d0, CONNECTING" module=grpc
time="2018-12-27T10:56:59.705356426Z" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc4201c44d0, READY" module=grpc
time="2018-12-27T10:56:59.705435674Z" level=info msg="Loading containers: start."
time="2018-12-27T10:56:59.754024850Z" level=warning msg="Running modprobe bridge br_netfilter failed with message: modprobe: WARNING: Module br_netfilter not found.\ninsmod /lib/modules/3.13.0-161-generic/kernel/net/llc/llc.ko \ninsmod /lib/modules/3.13.0-161-generic/kernel/net/802/stp.ko \ninsmod /lib/modules/3.13.0-161-generic/kernel/net/bridge/bridge.ko \n, error: exit status 1"
time="2018-12-27T10:57:00.125023434Z" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/16. Daemon option --bip can be used to set a preferred IP address"
time="2018-12-27T10:57:00.251582885Z" level=warning msg="Couldn't run auplink before unmount /var/lib/docker/aufs/mnt/8220418c28188d7cf01d3381990aa2c225c6a8ff35d7dd6d8b39453e470a1bc3" error="exec: \"auplink\": executable file not found in $PATH" storage-driver=aufs
time="2018-12-27T10:57:00.253372540Z" level=error msg="bc9ba169016d21bfed3776f825a9bb51ba8d9836be6f94a23a541b64ed0fe394 cleanup: failed to delete container from containerd: no such container"
time="2018-12-27T10:57:00.253400555Z" level=error msg="Failed to start container bc9ba169016d21bfed3776f825a9bb51ba8d9836be6f94a23a541b64ed0fe394: Cannot link to a non running container: /redis-server AS /telegraf/redis-server"
time="2018-12-27T10:57:00.256516394Z" level=info msg="Loading containers: done."
time="2018-12-27T10:57:00.356821716Z" level=info msg="Docker daemon" commit=e68fc7a graphdriver(s)=aufs version=18.06.1-ce
time="2018-12-27T10:57:00.357380275Z" level=info msg="Daemon has completed initialization"
time="2018-12-27T10:57:00.366963765Z" level=warning msg="Could not register builder git source: failed to find git binary: exec: \"git\": executable file not found in $PATH"
time="2018-12-27T10:57:00.379811565Z" level=info msg="API listen on /var/run/docker.sock"
time="2018-12-27T10:57:04.271532307Z" level=warning msg="Couldn't run auplink before unmount /var/lib/docker/aufs/mnt/91679f2d666bd39d307dce8d9de2632649bbbcff84b5835496ce3f93f7d2bbd0" error="exec: \"auplink\": executable file not found in $PATH" storage-driver=aufs
time="2018-12-27T10:57:04.275883676Z" level=error msg="Handler for POST /v1.38/containers/redis-server/start returned error: id already in use"
```

anything else?

**Comment by hashedvalue** [comment]
I am having similar problem with:
# docker version
Client:
 Version:           18.09.0
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        4d60db4
 Built:             Wed Nov  7 00:48:22 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.0
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       4d60db4
  Built:            Wed Nov  7 00:19:08 2018
  OS/Arch:          linux/amd64
  Experimental:     false

And:
# cat /etc/os-release
NAME="Red Hat Enterprise Linux Server"
VERSION="7.6 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.6"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.6 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.6:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.6
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.6"

Basically after OS restart part of the containers despite RestartPolicy set to always does not start with an error similar to:

_lut 06 07:26:01 server01 dockerd[8220]: time="2019-02-06T07:26:01.562495229+01:00" level=error msg="Failed to start container caf106a7eb2b8648e0c8edc5cca5c487dcd571da6976b6ed6c3c75061cb141ac: id already in use"_

This is problematic since every time a restart happens containers need to be restarted with some external mechanism, for example cron.
